complete_test.go

```diff
@@ -214,6 +223,42 @@ func TestComplete(t *testing.T) {
 	}
 }
 
+// ExampleComplete_outputCapturing demonstrates the ability to capture
+// the output of Complete() invocations, crucial for integration tests.
+func ExampleComplete_outputCapturing() {
+	// a stringLookup function maps one string to another.
+	// os.GetEnv is an instance of such a function.
+	type stringLookup = func(string) string
+
+	// promptEnv returns stringLookup func that emulates the environment
+	// variables a shell would set when its prompt has the given contents.
+	var promptEnv = func(contents string) stringLookup {
+		return func(key string) string {
+			switch key {
+			case "COMP_LINE":
+				return contents
+			case "COMP_POINT":
+				return strconv.Itoa(len(contents))
+			}
+			return ""
+		}
+	}
+
+	defer func(f func(int)) { exit = f }(exit)
+	defer func(f stringLookup) { getEnv = f }(getEnv)
```

Comment:
Maybe it would be a bit clearer just to copy the values into a local variable and just invoke a single defer function that restores them?

Also, I think it would make things easier to read if `exit` and `getEnv` one after the other.

Reply:
On one hand, I agree.

On the other... Each of these lines is a completely self-contained instance of that curious Go-specific "deferred self-restoring memento" pattern (I think of them as "time-capsules"). Things can be refactored, renamed, moved around, but each of those closures holds on to the respective original value and will faithfully restore it, once the enclosing scope is done, without littering the latter with variables that could accidentally get in the way of other stuff. Once you get used to this pattern, it reads very naturally and the intention is unambiguously clear — nothing else can happen than a variable getting back its original value as soon as the enclosing function returns.

<details><summary>At one point I considered merging both of those closures in a single one, but that felt markedly clumsier to me.</summary>

```diff
diff --git a/complete_test.go b/complete_test.go
index 182ca4c..e0cf61e 100644
--- a/complete_test.go
+++ b/complete_test.go
@@ -244,8 +244,10 @@ func ExampleComplete_outputCapturing() {
 		}
 	}
 
-	defer func(f func(int)) { exit = f }(exit)
-	defer func(f stringLookup) { getEnv = f }(getEnv)
+	defer func(_exit func(int), _getEnv stringLookup) {
+		exit = _exit
+		getEnv = _getEnv
+	}(exit, getEnv)
 	exit = func(int) {}
 
 	// This is where the actual example starts:
```
</details>

Either way, it's not like I'm surgically attached to those lines, if you have a clear idea of what you'd prefer to see, put it in a comment and I'll implement it precisely that way in the PR.

Reply:
<details><summary>Oh, right, I totally forgot about multiple assignment.</summary>

```diff
diff --git a/complete_test.go b/complete_test.go
index 182ca4c..9d82263 100644
--- a/complete_test.go
+++ b/complete_test.go
@@ -244,8 +244,9 @@ func ExampleComplete_outputCapturing() {
 		}
 	}
 
-	defer func(f func(int)) { exit = f }(exit)
-	defer func(f stringLookup) { getEnv = f }(getEnv)
+	defer func(_exit func(int), _getEnv stringLookup) {
+		exit, getEnv = _exit, _getEnv
+	}(exit, getEnv)
 	exit = func(int) {}
 
 	// This is where the actual example starts:
```

I'm still not sold though.
</details>

Was this what you had in mind?

Reply:
The current implementation is fine.
However, I don't understand why you explicitly need to pass them to the deferred function, and not borrow them from the outer scope:

```
prevExit, prevGetEnv := ...
defer func() {  exit, getEnv = prevExit, prevGetEnv }()
```

Reply:
<details><summary>No need, really, just a personal preference.</summary>

1. I don't have to come up with a variable name (or a naming convention) that's semantic yet unique in the enclosing (outer) scope and has no potential for collisions as the code grows and changes with time.
2. There is absolutely zero chance of accidentally messing things up (e.g. overwriting the old value).
3. It makes for a single, clear, concise, perfectly reproducible pattern to follow with no additional thought to be given to, ever again.
4. Refactoring is straightforward, no diff churn — every single `defer` line stands on its own.

So, instead of

```golang
// Would "oldX" be better? Or "originalX"? What about any other prefix?
prevFoo := somepkg.Foo
prevBar := somepkg.Bar
prevQuux := somepkg.Quux
prevGarply := somepkg.Garply
prevCorge := somepkg.Corge
defer func() {
	somepkg.Foo = prevFoo
	somepkg.Bar = prevBar
	somepkg.Quux = prevQuux
	somepkg.Garply = prevGarply
	somepkg.Corge = prevCorge
}()
```

or

```golang
prevFoo, prevBar, prevQuux := somepkg.Foo, somepkg.Bar, somepkg.Quux
prevGarply, prevCorge := somepkg.Garply, somepkg.Corge
defer func() {
	somepkg.Foo, somepkg.Bar, somepkg.Quux = prevFoo, prevBar
	somepkg.Garply, somepkg.Corge = prevQuux, prevGarply, prevCorge
}()
```

or any other variation you could come up with, I would always have

```golang
defer func(v Footype) { somepkg.Foo = v }(somepkg.Foo)
defer func(v Bartype) { somepkg.Bar = v }(somepkg.Bar)
defer func(v Quuxtype) { somepkg.Quux = v }(somepkg.Quux)
defer func(v Garplytype) { somepkg.Garply = v }(somepkg.Garply)
defer func(v Corgetype) { somepkg.Corge = v }(somepkg.Corge)
```

The fact that I chose `v` for the parameter name bears no significance from closure to closure, I could rename it in any or every one of them to any other valid symbol name, that's a detail that should not take up any brain cycles for you, me or anyone.

It also looks cleaner to me than any variation with multiple assignment.
</details>

Alright then, your final word, is this:

```diff
diff --git a/complete_test.go b/complete_test.go
index 9870910..3454926 100644
--- a/complete_test.go
+++ b/complete_test.go
@@ -226,8 +226,8 @@ func TestComplete(t *testing.T) {
 // ExampleComplete_outputCapturing demonstrates the ability to capture
 // the output of Complete() invocations, crucial for integration tests.
 func ExampleComplete_outputCapturing() {
-	defer func(f func(int)) { exit = f }(exit)
-	defer func(f getEnvFn) { getEnv = f }(getEnv)
+	prevExit, prevGetEnv := exit, getEnv
+	defer func() { exit, getEnv = prevExit, prevGetEnv }()
 	exit = func(int) {}
 
 	// This is where the actual example starts:
```

how you want it?